### PR TITLE
Preserve ReplyToAddress for ingested errors

### LINKED ISSUE #911: Upon retrying messages, ServiceControl is changing the ReplyTo address of messages incorrectly
User reported:
https://groups.google.com/forum/?pli=1#!topic/particularsoftware/CnObyrIkuK0

Steps to Repro:
- Setup 2 machines PC1 and PC2
- Install SC on PC1 and run the Full Duplex Sample's Server on PC1 (tweak it to throw an exception)
- Run the Full Duplex Sample's Client on PC2 and send a message 
- Close the Server App on PC2
- Retry the failed message either using SP or SI
- Inspect the message that's on the Server's queue and look at the ReplyTo Address and notice that it is using the SC's machine name which is PC1 and not PC2.

### PR DESCRIPTION
**Pull request by mikeminutillo**
Fixes #911

### CODE REVIEW

**Comment by mikeminutillo** [comment]
@Particular/servicecontrol-maintainers please review and keep an eye on the build status for this one.

**Comment by indualagarsamy** [comment]
@mikeminutillo - what about here? 
https://github.com/Particular/ServiceControl/blob/preserve-replytoaddress/src/ServiceControl/Infrastructure/TransportMessageExtensions.cs#L31

Is this relevant here as well?

**Comment by indualagarsamy** [comment]
@mikeminutillo - do we need any changes to this test here, based on the bug and now?
https://github.com/Particular/ServiceControl/blob/preserve-replytoaddress/src/ServiceControl.UnitTests/Infrastructure/TransportMessageExtensionsTests.cs#L51

**Comment by mikeminutillo** [comment]
This won't fix data on disk. It will only work for newly ingested Failed Messages. 

I don't know why we are overwriting the ReplyToAddress header that was ingested with the metadata one at all. Removing [the code that does this](https://github.com/Particular/ServiceControl/blob/master/src/ServiceControl/Recoverability/Retries/RetryProcessor.cs#L199) will fix it but I am uncertain why this code was present in the first place. Investigating further.

**Comment by mikeminutillo** [comment]
This appears to be a holdout from when SC ran on NSB4. In that version, ReplyToAddress was a property of TransportMessage. The reason to pull it out as a separate piece of metadata was because in V4, `ReplyToAddress` was [just a property and was not linked to the header](https://github.com/Particular/NServiceBus/blob/4.6.10/src/NServiceBus.Core/TransportMessage.cs#L111).

**Comment by mikeminutillo** [comment]
The test referenced is a different path of code entirely. It uses the headers directly and is not supposed to preserve the machine name. It is looking for a Logical Endpoint name so it explicitly drops the machine name from the physical address. 

Having said that, the test should include a machine name to prove that it's getting dropped.

**Comment by mikeminutillo** [comment]
This fix has a green build. 

Should we remove the metadata `ReplyToAddress` on Processing Attempts? I don't believe anyone is using it. It doesn't have to be a part of this fix. We can add a refactoring issue to the backlog.

Also should I revert the change to the Ingestion pipe to put it back the way it was for now? As I said, I'm pretty sure there is no code using it.

**Comment by mikeminutillo** [comment]
This symptom is still present when running on the SQL Server transport but it is a [different issue](https://github.com/Particular/NServiceBus.SqlServer/issues/345)

File: src/ServiceControl/Operations/ErrorQueueImport.cs [hunk 1]
```diff
@@ -137,14 +137,16 @@ private void InnerHandle(TransportMessage message)
 
             var failureDetails = failedMessageFactory.ParseFailureDetails(message.Headers);
 
+            var replyToAddress = message.ReplyToAddress == null ? null : message.ReplyToAddress.ToString();
```
Comment (SzymonPobiega) [inline src/ServiceControl/Operations/ErrorQueueImport.cs]: Given the fact that `ReplyToAddress` is implemented as

```
get
      {
        string destination;
        if (this.Headers.TryGetValue("NServiceBus.ReplyToAddress", out destination))
          return Address.Parse(destination);
        return (Address) null;
      }
```
we should just plain get the header value here and not pass it through the `Address` class. This way we'll preserve whatever the transport-defined value was in the first place.

**Comment by SzymonPobiega** [comment]
@mikeminutillo I believe that if we change this to what I proposed, we won't need the SQL server fix.

File: src/ServiceControl/Operations/ErrorQueueImport.cs (hunk 1 above)
Comment (mikeminutillo) [inline src/ServiceControl/Operations/ErrorQueueImport.cs]: You can safely ignore this value. It used to be used to preserve the value when the underlying transport core did not do so. It should eventually be removed.

The main issue is [here](https://github.com/Particular/ServiceControl/blob/d78e169da41f38e54645b254465a40dcd0ad6d09/src/ServiceControl/Recoverability/Retries/RetryProcessor.cs#L211). This transport message contains the unadulterated ReplyToAddress header but the transport will run it through `Address.Parse().ToString()` before writing it to the table as outlined in https://github.com/Particular/NServiceBus.SqlServer/issues/345

**Comment by mikeminutillo** [comment]
If a message has already been through a retry attempt (via ServiceControl), and the ReplyToAddress has been corrupted, if the processing fails again the the newly imported ReplyToAddress header will be recorded in it's corrupt state. 

I just added code to detect and fix this case (at least for MSMQ) on outgoing retries. I extracted the behavior into a new class and unit tested it. In order to acceptance test it, we need the sending endpoint to run on a different machine than ServiceControl runs on :s

**Comment by boblangley** [comment]
Merged #912

**Pull request merged**